### DISCUSSION TITLE
Environment variables not available on Vercel

### DISCUSSION THREAD

**Discussion by foxted** [vercel/turborepo#8627] [Help] [answered]
### Summary

Recently migrated a manual monorepo to turbo repo, and when deploying, my environment variables aren't detected. 

I have tried passing the `env`, `globalEnv` & `globalPassThroughEnv` to the main turbo.json, didn't work. Tried the same but splitting the turbo.json into each app folder, didn't work. 

I can build the project successfully locally, but when deploying to Vercel, I keep getting a message (in this case from supabase-js) that my environment variable is missing. 

Any tips on getting this to work? I like the idea of the strict mode, but I'm out of ideas on how to make this work. 

### Additional information

```js
# turbo.json
{
  "$schema": "https://turbo.build/schema.json",
  "tasks": {
    "build": {
      "dependsOn": ["^build"],
      "outputs": [".next/**", "!.next/cache/**", "dist/**"]
    },
    "lint": {
      "dependsOn": ["^lint"]
    },
    "dev": {
      "persistent": true,
      "cache": false
    },
    "webhook:listen": {
      "dependsOn": ["^webhook:listen"],
      "persistent": true
    },
    "check-types": {
      "dependsOn": ["^check-types"]
    }
  }
}

# apps/app-1/turbo.json
{
  "extends": ["//"],
  "globalPassThroughEnv": ["SUPABASE_URL"]
}

# apps/app-2/turbo.json
{
  "extends": ["//"],
  "globalPassThroughEnv": ["SUPABASE_URL"]
}
```
```


### Example

_No response_

**Comment by foxted** [comment, marked as answer]
Found the answer: 

In Vercel, had to update the build command to `cd ../.. && turbo run build --filter=app-1` and use `passThroughEnv` in my `turbo.json` for each apps.

**Reply by MarkAtOmniux** [reply]
would you mind pasting a copy of your turbo.json and your package.json for future searchers please?

**Reply by foxted** [reply]
Main turbo.json:

```json
{
  "$schema": "https://turbo.build/schema.json",
  "tasks": {
    "build": {
      "dependsOn": ["^build"],
      "inputs": [
        "$TURBO_DEFAULT$",
        ".env.production.local",
        ".env.local",
        ".env.production",
        ".env"
      ],
      "outputs": [".next/**", "!.next/cache/**", "dist/**"]
    },
    "lint": {
      "dependsOn": ["^lint"]
    },
    "dev": {
      "persistent": true,
      "cache": false,
      "inputs": [
        "$TURBO_DEFAULT$",
        ".env.development.local",
        ".env.local",
        ".env.development",
        ".env"
      ]
    },
    "webhook:listen": {
      "dependsOn": ["^webhook:listen"],
      "persistent": true
    },
    "check-types": {
      "dependsOn": ["^check-types"]
    }
  }
}
```

app1/turbo.json

```json
{
  "extends": ["//"],
  "tasks": {
    "build": {
      "passThroughEnv": ["CLERK_SECRET_KEY", "CLERK_WEBHOOK_SECRET", "SUPABASE_URL", "SUPABASE_SERVICE_ROLE_KEY"]
    }
  }
}
```

app2/turbo.json

```json
{
  "extends": ["//"],
  "tasks": {
    "build": {
      "passThroughEnv": ["CONVERT_KIT_API_KEY", "CONVERT_KIT_API_SECRET"]
    }
  }
}
```

The build command was overwritten directly in Vercel build settings.

**Reply by clearly-outsane** [reply]
And where were the .envs? was there one in the root of the monorepo or one in each app and package folder?

**Comment by kane1525** [comment]
explanation why this happens: turbo by default run in strict mode and it filters out all env variables that not specified at config

passThroughEnv is not a best practise as it simply removes these variables from creating cache hashes while in most cases it's essential for correct turbo's cache work.

Here is a cleaner solution:

at root turbo.json:

```
	"build": {
			"dependsOn": ["^build"],
			"inputs": [
				"$TURBO_DEFAULT$",
				".env.production.local",
				".env.local",
				".env.production",
				".env"
			  ],
			"outputs": [".next/**", "!.next/cache/**", "dist/**"]
		},
```

and apps/project/turbo.json:

```
{
    "extends": ["//"],
    "tasks": {
      "build": {
        "env": [
            "ENV_1",
             "ENV_2"
        ]
      }
    }
  }
```
